File: packages/python/openpkpd/trial/population.py
"""
OpenPKPD Virtual Population Generation

Virtual population generation for clinical trial simulation including:
- Demographic specifications
- Disease state modeling
- Covariate correlations
"""

from __future__ import annotations

from dataclasses import dataclass, field
from typing import Any, Dict, List, Optional, Tuple
import random
import math


@dataclass
class DemographicSpec:
    """
    Demographic distribution specification.

    Attributes:
        age_mean: Mean age (years)
        age_sd: Age standard deviation
        age_range: Age range limits (min, max)
        weight_mean: Mean weight (kg)
        weight_sd: Weight standard deviation
        weight_range: Weight range limits
        female_proportion: Proportion of females (0-1)
        race_distribution: Race/ethnicity distribution

    Example:
        >>> spec = DemographicSpec(age_mean=45.0, age_sd=12.0)
    """
    age_mean: float = 45.0
    age_sd: float = 15.0
    age_range: Tuple[float, float] = (18.0, 75.0)
    weight_mean: float = 75.0
    weight_sd: float = 15.0
    weight_range: Tuple[float, float] = (40.0, 150.0)
    female_proportion: float = 0.5
    race_distribution: Dict[str, float] = field(default_factory=lambda: {
        "caucasian": 0.70,
        "asian": 0.15,
        "black": 0.10,
        "other": 0.05,
    })


@dataclass
class DiseaseSpec:
    """
    Disease state specification.

    Attributes:
        name: Disease name
        severity_distribution: Distribution of severity levels
        baseline_biomarker_mean: Mean baseline biomarker value
        baseline_biomarker_sd: Baseline biomarker standard deviation

    Example:
        >>> spec = DiseaseSpec(
        ...     name="diabetes",
        ...     severity_distribution={"mild": 0.3, "moderate": 0.5, "severe": 0.2}
        ... )
    """
    name: str
    severity_distribution: Dict[str, float] = field(default_factory=lambda: {
        "mild": 0.30,
        "moderate": 0.50,
        "severe": 0.20,
    })
    baseline_biomarker_mean: float = 100.0
    baseline_biomarker_sd: float = 25.0


@dataclass
class VirtualPopulationSpec:
    """
    Specification for virtual population generation.

    Attributes:
        demographics: Demographic distribution
        disease: Disease specification (optional)
        covariate_correlations: Covariate correlations
        seed: Random seed for reproducibility

    Example:
        >>> spec = VirtualPopulationSpec(
        ...     demographics=DemographicSpec(age_mean=50.0),
        ...     seed=42
        ... )
    """
    demographics: DemographicSpec = field(default_factory=DemographicSpec)
    disease: Optional[DiseaseSpec] = None
    covariate_correlations: Dict[Tuple[str, str], float] = field(default_factory=dict)
    seed: int = 12345


@dataclass
class VirtualSubject:
    """
    Virtual subject for trial simulation.

    Attributes:
        id: Subject ID
        age: Age (years)
        weight: Weight (kg)
        sex: Sex ('male' or 'female')
        race: Race/ethnicity
        disease_severity: Disease severity level (optional)
        baseline_biomarker: Baseline biomarker value (optional)
        other: Other covariates

    Example:
        >>> subject = VirtualSubject(
        ...     id=1, age=45.0, weight=75.0, sex='male', race='caucasian'
        ... )
    """
    id: int
    age: float
    weight: float
    sex: str
    race: str
    disease_severity: Optional[str] = None
    baseline_biomarker: Optional[float] = None
    other: Dict[str, float] = field(default_factory=dict)

    def to_dict(self) -> Dict[str, Any]:
        """Convert to dictionary."""
        return {
            "id": self.id,
            "age": self.age,
            "weight": self.weight,
            "sex": self.sex,
            "race": self.race,
            "disease_severity": self.disease_severity,
            "baseline_biomarker": self.baseline_biomarker,
            "other": self.other,
        }


def _sample_truncated_normal(
    rng: random.Random,
    mean: float,
    sd: float,
    lower: float,
    upper: float,
) -> float:
    """Sample from a truncated normal distribution."""
    for _ in range(1000):
        x = rng.gauss(mean, sd)
        if lower <= x <= upper:
            return x
    return max(lower, min(upper, rng.gauss(mean, sd)))


def _sample_categorical(rng: random.Random, distribution: Dict[str, float]) -> str:
    """Sample from a categorical distribution."""
    u = rng.random()
    cumsum = 0.0
    for category, prob in distribution.items():
        cumsum += prob
        if u <= cumsum:
            return category
    return list(distribution.keys())[-1]


def default_demographic_spec() -> DemographicSpec:
    """
    Create default demographic specification for healthy volunteers.

    Returns:
        DemographicSpec object

    Example:
        >>> spec = default_demographic_spec()
    """
    return DemographicSpec(
        age_mean=35.0,
        age_sd=10.0,
        age_range=(18.0, 55.0),
        weight_mean=75.0,
        weight_sd=12.0,
        weight_range=(50.0, 120.0),
        female_proportion=0.5,
    )


def healthy_volunteer_spec() -> DemographicSpec:
    """
    Create demographic specification for healthy volunteer studies.

    Returns:
        DemographicSpec object

    Example:
        >>> spec = healthy_volunteer_spec()
    """
    return DemographicSpec(
        age_mean=30.0,
        age_sd=8.0,
        age_range=(18.0, 45.0),
        weight_mean=72.0,
        weight_sd=10.0,
        weight_range=(55.0, 100.0),
        female_proportion=0.5,
        race_distribution={
            "caucasian": 0.65,
            "asian": 0.20,
            "black": 0.10,
            "other": 0.05,
        },
    )


def patient_population_spec(
    disease: str,
    severity_weights: Optional[Dict[str, float]] = None,
) -> Tuple[DemographicSpec, DiseaseSpec]:
    """
    Create demographic and disease specification for patient population.

    Args:
        disease: Disease type ('diabetes', 'hepatic', 'renal', 'oncology')
        severity_weights: Custom severity distribution

    Returns:
        Tuple of (DemographicSpec, DiseaseSpec)

    Example:
        >>> demo, dis = patient_population_spec('diabetes')
    """
    if disease == "diabetes":
        demo = DemographicSpec(
            age_mean=55.0, age_sd=12.0, age_range=(30.0, 75.0),
            weight_mean=90.0, weight_sd=18.0, weight_range=(60.0, 150.0),
            female_proportion=0.45,
            race_distribution={"caucasian": 0.60, "asian": 0.15, "black": 0.15, "hispanic": 0.10},
        )
        severity = severity_weights or {"mild": 0.25, "moderate": 0.50, "severe": 0.25}
        dis = DiseaseSpec(
            name="diabetes",
            severity_distribution=severity,
            baseline_biomarker_mean=8.5,  # HbA1c
            baseline_biomarker_sd=1.5,
        )

    elif disease == "hepatic":
        demo = DemographicSpec(
            age_mean=52.0, age_sd=14.0, age_range=(25.0, 75.0),
            weight_mean=80.0, weight_sd=15.0, weight_range=(50.0, 130.0),
            female_proportion=0.40,
        )
        severity = severity_weights or {"mild": 0.40, "moderate": 0.40, "severe": 0.20}
        dis = DiseaseSpec(
            name="hepatic_impairment",
            severity_distribution=severity,
            baseline_biomarker_mean=2.5,
            baseline_biomarker_sd=0.8,
        )

    elif disease == "renal":
        demo = DemographicSpec(
            age_mean=58.0, age_sd=15.0, age_range=(25.0, 80.0),
            weight_mean=78.0, weight_sd=16.0, weight_range=(45.0, 130.0),
            female_proportion=0.45,
            race_distribution={"caucasian": 0.60, "asian": 0.15, "black": 0.20, "other": 0.05},
        )
        severity = severity_weights or {"mild": 0.30, "moderate": 0.40, "severe": 0.20, "esrd": 0.10}
        dis = DiseaseSpec(
            name="renal_impairment",
            severity_distribution=severity,
            baseline_biomarker_mean=45.0,  # eGFR
            baseline_biomarker_sd=20.0,
        )

    elif disease == "oncology":
        demo = DemographicSpec(
            age_mean=62.0, age_sd=12.0, age_range=(25.0, 85.0),
            weight_mean=72.0, weight_sd=15.0, weight_range=(40.0, 120.0),
            female_proportion=0.48,
        )
        severity = severity_weights or {"stage_I": 0.15, "stage_II": 0.25, "stage_III": 0.35, "stage_IV": 0.25}
        dis = DiseaseSpec(
            name="cancer",
            severity_distribution=severity,
            baseline_biomarker_mean=50.0,
            baseline_biomarker_sd=30.0,
        )

    else:
        demo = DemographicSpec(
            age_mean=50.0, age_sd=15.0, age_range=(18.0, 80.0),
            weight_mean=78.0, weight_sd=15.0, weight_range=(45.0, 140.0),
            female_proportion=0.50,
        )
        severity = severity_weights or {"mild": 0.30, "moderate": 0.50, "severe": 0.20}
        dis = DiseaseSpec(
            name=disease,
            severity_distribution=severity,
            baseline_biomarker_mean=100.0,
            baseline_biomarker_sd=25.0,
        )

    return demo, dis


def generate_virtual_population(
    n: int,
    spec: Optional[VirtualPopulationSpec] = None,
    demographics: Optional[DemographicSpec] = None,
    disease: Optional[DiseaseSpec] = None,
    seed: Optional[int] = None,
) -> List[VirtualSubject]:
    """
    Generate a virtual population.

    Args:
        n: Number of subjects
        spec: Complete population specification (overrides other args)
        demographics: Demographic specification
        disease: Disease specification
        seed: Random seed

    Returns:
        List of VirtualSubject objects

    Example:
        >>> pop = generate_virtual_population(100)
        >>> pop = generate_virtual_population(100, seed=42)
        >>> pop = generate_virtual_population(
        ...     100,
        ...     demographics=DemographicSpec(age_mean=50.0)
        ... )
    """
    if spec is not None:
        demographics = spec.demographics
        disease = spec.disease
        seed = spec.seed

    if demographics is None:
        demographics = DemographicSpec()

    if seed is None:
        seed = 12345

    rng = random.Random(seed)
    population = []

    for i in range(n):
        age = _sample_truncated_normal(
            rng, demographics.age_mean, demographics.age_sd,
            demographics.age_range[0], demographics.age_range[1]
        )
        weight = _sample_truncated_normal(
            rng, demographics.weight_mean, demographics.weight_sd,
            demographics.weight_range[0], demographics.weight_range[1]
        )
        sex = "female" if rng.random() < demographics.female_proportion else "male"
        race = _sample_categorical(rng, demographics.race_distribution)

        disease_severity = None
        baseline_biomarker = None

        if disease is not None:
            disease_severity = _sample_categorical(rng, disease.severity_distribution)
            baseline_biomarker = _sample_truncated_normal(
                rng,
                disease.baseline_biomarker_mean,
                disease.baseline_biomarker_sd,
                max(0.0, disease.baseline_biomarker_mean - 3 * disease.baseline_biomarker_sd),
                disease.baseline_biomarker_mean + 3 * disease.baseline_biomarker_sd,
            )

        subject = VirtualSubject(
            id=i + 1,
            age=age,
            weight=weight,
            sex=sex,
            race=race,
            disease_severity=disease_severity,
            baseline_biomarker=baseline_biomarker,
        )
        population.append(subject)

    return population


def summarize_population(population: List[VirtualSubject]) -> Dict[str, Any]:
    """
    Generate summary statistics for a virtual population.

    Args:
        population: List of VirtualSubject objects

    Returns:
        Dictionary of summary statistics

    Example:
        >>> pop = generate_virtual_population(100)
        >>> summary = summarize_population(pop)
        >>> print(f"Mean age: {summary['age_mean']:.1f}")
    """
    n = len(population)
    if n == 0:
        return {"n": 0}

    ages = [s.age for s in population]
    weights = [s.weight for s in population]
    n_female = sum(1 for s in population if s.sex == "female")

    # Race counts
    race_counts: Dict[str, int] = {}
    for s in population:
        race_counts[s.race] = race_counts.get(s.race, 0) + 1

    # Disease severity counts
    severity_counts: Dict[str, int] = {}
    has_disease = any(s.disease_severity is not None for s in population)
    if has_disease:
        for s in population:
            if s.disease_severity is not None:
                severity_counts[s.disease_severity] = severity_counts.get(s.disease_severity, 0) + 1

    age_mean = sum(ages) / n
    weight_mean = sum(weights) / n

    summary = {
        "n": n,
        "age_mean": age_mean,
        "age_sd": math.sqrt(sum((a - age_mean) ** 2 for a in ages) / (n - 1)) if n > 1 else 0.0,
        "age_range": (min(ages), max(ages)),
        "weight_mean": weight_mean,
        "weight_sd": math.sqrt(sum((w - weight_mean) ** 2 for w in weights) / (n - 1)) if n > 1 else 0.0,
        "weight_range": (min(weights), max(weights)),
        "female_proportion": n_female / n,
        "race_distribution": {k: v / n for k, v in race_counts.items()},
    }

    if has_disease:
        biomarkers = [s.baseline_biomarker for s in population if s.baseline_biomarker is not None]
        if biomarkers:
            bm_mean = sum(biomarkers) / len(biomarkers)
            summary["biomarker_mean"] = bm_mean
            summary["biomarker_sd"] = math.sqrt(
                sum((b - bm_mean) ** 2 for b in biomarkers) / (len(biomarkers) - 1)
            ) if len(biomarkers) > 1 else 0.0
        summary["severity_distribution"] = {k: v / n for k, v in severity_counts.items()}

    return summary

File: packages/core/src/trial/population.jl
# Virtual Population Generation
# Functions for generating virtual subject populations

export generate_virtual_population, sample_demographics, apply_covariate_correlations
export default_demographic_spec, healthy_volunteer_spec, patient_population_spec
export summarize_population

using StableRNGs
using Random

"""
    default_demographic_spec()

Create default demographic specification for healthy volunteers.

# Returns
- `DemographicSpec`: Default demographic specification
"""
function default_demographic_spec()
    return DemographicSpec(
        age_mean = 35.0,
        age_sd = 10.0,
        age_range = (18.0, 55.0),
        weight_mean = 75.0,
        weight_sd = 12.0,
        weight_range = (50.0, 120.0),
        female_proportion = 0.5,
        race_distribution = Dict(:caucasian => 0.70, :asian => 0.15, :black => 0.10, :other => 0.05)
    )
end


"""
    healthy_volunteer_spec()

Create demographic specification for healthy volunteer studies.

# Returns
- `DemographicSpec`: Healthy volunteer specification
"""
function healthy_volunteer_spec()
    return DemographicSpec(
        age_mean = 30.0,
        age_sd = 8.0,
        age_range = (18.0, 45.0),
        weight_mean = 72.0,
        weight_sd = 10.0,
        weight_range = (55.0, 100.0),
        female_proportion = 0.5,
        race_distribution = Dict(:caucasian => 0.65, :asian => 0.20, :black => 0.10, :other => 0.05)
    )
end


"""
    patient_population_spec(disease::Symbol; severity_weights=nothing)

Create demographic specification for patient population.

# Arguments
- `disease::Symbol`: Disease type (:diabetes, :hepatic, :renal, :cardiac, :oncology)

# Keyword Arguments
- `severity_weights::Dict{Symbol, Float64}`: Custom severity distribution

# Returns
- `Tuple{DemographicSpec, DiseaseSpec}`: Demographics and disease specification
"""
function patient_population_spec(disease::Symbol;
        severity_weights::Union{Nothing, Dict{Symbol, Float64}} = nothing)

    if disease == :diabetes
        demo = DemographicSpec(
            age_mean = 55.0, age_sd = 12.0, age_range = (30.0, 75.0),
            weight_mean = 90.0, weight_sd = 18.0, weight_range = (60.0, 150.0),
            female_proportion = 0.45,
            race_distribution = Dict(:caucasian => 0.60, :asian => 0.15, :black => 0.15, :hispanic => 0.10)
        )
        severity = severity_weights !== nothing ? severity_weights :
            Dict(:mild => 0.25, :moderate => 0.50, :severe => 0.25)
        dis = DiseaseSpec(:diabetes;
            severity_distribution = severity,
            baseline_biomarker_mean = 8.5,  # HbA1c
            baseline_biomarker_sd = 1.5)

    elseif disease == :hepatic
        demo = DemographicSpec(
            age_mean = 52.0, age_sd = 14.0, age_range = (25.0, 75.0),
            weight_mean = 80.0, weight_sd = 15.0, weight_range = (50.0, 130.0),
            female_proportion = 0.40,
            race_distribution = Dict(:caucasian => 0.65, :asian => 0.20, :black => 0.10, :other => 0.05)
        )
        severity = severity_weights !== nothing ? severity_weights :
            Dict(:mild => 0.40, :moderate => 0.40, :severe => 0.20)
        dis = DiseaseSpec(:hepatic_impairment;
            severity_distribution = severity,
            baseline_biomarker_mean = 2.5,  # Child-Pugh score component
            baseline_biomarker_sd = 0.8)

    elseif disease == :renal
        demo = DemographicSpec(
            age_mean = 58.0, age_sd = 15.0, age_range = (25.0, 80.0),
            weight_mean = 78.0, weight_sd = 16.0, weight_range = (45.0, 130.0),
            female_proportion = 0.45,
            race_distribution = Dict(:caucasian => 0.60, :asian => 0.15, :black => 0.20, :other => 0.05)
        )
        severity = severity_weights !== nothing ? severity_weights :
            Dict(:mild => 0.30, :moderate => 0.40, :severe => 0.20, :esrd => 0.10)
        dis = DiseaseSpec(:renal_impairment;
            severity_distribution = severity,
            baseline_biomarker_mean = 45.0,  # eGFR
            baseline_biomarker_sd = 20.0)

    elseif disease == :oncology
        demo = DemographicSpec(
            age_mean = 62.0, age_sd = 12.0, age_range = (25.0, 85.0),
            weight_mean = 72.0, weight_sd = 15.0, weight_range = (40.0, 120.0),
            female_proportion = 0.48,
            race_distribution = Dict(:caucasian => 0.70, :asian => 0.12, :black => 0.12, :other => 0.06)
        )
        severity = severity_weights !== nothing ? severity_weights :
            Dict(:stage_I => 0.15, :stage_II => 0.25, :stage_III => 0.35, :stage_IV => 0.25)
        dis = DiseaseSpec(:cancer;
            severity_distribution = severity,
            baseline_biomarker_mean = 50.0,  # Tumor burden
            baseline_biomarker_sd = 30.0)

    else
        # Default generic patient population
        demo = DemographicSpec(
            age_mean = 50.0, age_sd = 15.0, age_range = (18.0, 80.0),
            weight_mean = 78.0, weight_sd = 15.0, weight_range = (45.0, 140.0),
            female_proportion = 0.50,
            race_distribution = Dict(:caucasian => 0.65, :asian => 0.15, :black => 0.15, :other => 0.05)
        )
        severity = severity_weights !== nothing ? severity_weights :
            Dict(:mild => 0.30, :moderate => 0.50, :severe => 0.20)
        dis = DiseaseSpec(disease;
            severity_distribution = severity,
            baseline_biomarker_mean = 100.0,
            baseline_biomarker_sd = 25.0)
    end

    return (demo, dis)
end


"""
    sample_truncated_normal(rng, mean, sd, lower, upper)

Sample from a truncated normal distribution using rejection sampling.
"""
function sample_truncated_normal(rng::AbstractRNG, mean::Float64, sd::Float64,
                                  lower::Float64, upper::Float64)
    # Simple rejection sampling for truncated normal
    max_attempts = 1000
    for _ in 1:max_attempts
        x = mean + sd * randn(rng)
        if lower <= x <= upper
            return x
        end
    end
    # Fallback: clamp to bounds
    return clamp(mean + sd * randn(rng), lower, upper)
end


"""
    sample_categorical(rng, distribution)

Sample from a categorical distribution.
"""
function sample_categorical(rng::AbstractRNG, distribution::Dict{Symbol, Float64})
    u = rand(rng)
    cumsum = 0.0
    for (category, prob) in distribution
        cumsum += prob
        if u <= cumsum
            return category
        end
    end
    # Fallback to last category
    return last(keys(distribution))
end


"""
    sample_demographics(spec::DemographicSpec, id::Int; rng=nothing)

Sample a single individual's demographics.

# Arguments
- `spec::DemographicSpec`: Demographic specification
- `id::Int`: Subject ID

# Keyword Arguments
- `rng::AbstractRNG`: Random number generator

# Returns
- `VirtualSubject`: Individual's demographics
"""
function sample_demographics(spec::DemographicSpec, id::Int;
                              rng::Union{Nothing, AbstractRNG} = nothing)
    if rng === nothing
        rng = StableRNG(12345 + id)
    end

    age = sample_truncated_normal(rng, spec.age_mean, spec.age_sd,
                                   spec.age_range[1], spec.age_range[2])
    weight = sample_truncated_normal(rng, spec.weight_mean, spec.weight_sd,
                                      spec.weight_range[1], spec.weight_range[2])
    sex = rand(rng) < spec.female_proportion ? :female : :male
    race = sample_categorical(rng, spec.race_distribution)

    return VirtualSubject(id, age, weight, sex, race, nothing, nothing, Dict{Symbol, Float64}())
end


"""
    apply_disease_characteristics(individual, disease_spec; rng=nothing)

Apply disease-specific characteristics to an individual.

# Arguments
- `individual::VirtualSubject`: Individual to modify
- `disease_spec::DiseaseSpec`: Disease specification

# Keyword Arguments
- `rng::AbstractRNG`: Random number generator

# Returns
- `VirtualSubject`: Individual with disease characteristics
"""
function apply_disease_characteristics(individual::VirtualSubject,
                                        disease_spec::DiseaseSpec;
                                        rng::Union{Nothing, AbstractRNG} = nothing)
    if rng === nothing
        rng = StableRNG(12345 + individual.id)
    end

    severity = sample_categorical(rng, disease_spec.severity_distribution)
    baseline_biomarker = sample_truncated_normal(rng,
        disease_spec.baseline_biomarker_mean,
        disease_spec.baseline_biomarker_sd,
        max(0.0, disease_spec.baseline_biomarker_mean - 3 * disease_spec.baseline_biomarker_sd),
        disease_spec.baseline_biomarker_mean + 3 * disease_spec.baseline_biomarker_sd)

    return VirtualSubject(
        individual.id,
        individual.age,
        individual.weight,
        individual.sex,
        individual.race,
        severity,
        baseline_biomarker,
        individual.other
    )
end


"""
    apply_covariate_correlations(individuals, correlations; rng=nothing)

Apply covariate correlations to a population.

Uses Gaussian copula approach for correlated covariates.

# Arguments
- `individuals::Vector{VirtualSubject}`: Population
- `correlations::Dict{Tuple{Symbol, Symbol}, Float64}`: Correlation specifications

# Keyword Arguments
- `rng::AbstractRNG`: Random number generator

# Returns
- `Vector{VirtualSubject}`: Population with correlated covariates
"""
function apply_covariate_correlations(individuals::Vector{VirtualSubject},
                                       correlations::Dict{Tuple{Symbol, Symbol}, Float64};
                                       rng::Union{Nothing, AbstractRNG} = nothing)
    if isempty(correlations)
        return individuals
    end

    if rng === nothing
        rng = StableRNG(12345)
    end

    n = length(individuals)

    # For now, apply a simple rank-based correlation adjustment
    # This is a simplified implementation - full Gaussian copula would be more complex

    for ((cov1, cov2), target_corr) in correlations
        if cov1 == :age && cov2 == :weight
            # Sort by age and apply partial reordering to create correlation
            ages = [ind.age for ind in individuals]
            weights = [ind.weight for ind in individuals]

            age_order = sortperm(ages)
            weight_order = sortperm(weights)

            # Mix orderings to achieve approximate target correlation
            blend = abs(target_corr)
            for i in 1:n
                if rand(rng) < blend
                    # Swap to match correlation
                    if target_corr > 0
                        # Positive correlation: high age -> high weight
                        # Already somewhat handled by demographic generation
                    else
                        # Negative correlation: high age -> low weight
                        # Would need to implement swapping logic
                    end
                end
            end
        end
    end

    return individuals
end


"""
    generate_virtual_population(spec::VirtualPopulationSpec, n::Int)

Generate a virtual population.

# Arguments
- `spec::VirtualPopulationSpec`: Population specification
- `n::Int`: Number of subjects

# Returns
- `Vector{VirtualSubject}`: Generated population

# Example
```julia
spec = VirtualPopulationSpec(
    demographics = DemographicSpec(age_mean=45.0, age_sd=12.0),
    seed = UInt64(42)
)
population = generate_virtual_population(spec, 100)
```
"""
function generate_virtual_population(spec::VirtualPopulationSpec, n::Int)
    rng = StableRNG(spec.seed)

    individuals = Vector{VirtualSubject}(undef, n)

    for i in 1:n
        ind = sample_demographics(spec.demographics, i; rng = rng)

        if spec.disease !== nothing
            ind = apply_disease_characteristics(ind, spec.disease; rng = rng)
        end

        individuals[i] = ind
    end

    # Apply covariate correlations
    individuals = apply_covariate_correlations(individuals,
                                                spec.covariate_correlations;
                                                rng = rng)

    return individuals
end


"""
    summarize_population(population::Vector{VirtualSubject})

Generate summary statistics for a virtual population.

# Arguments
- `population::Vector{VirtualSubject}`: Population

# Returns
- `Dict{Symbol, Any}`: Summary statistics
"""
function summarize_population(population::Vector{VirtualSubject})
    n = length(population)

    ages = [ind.age for ind in population]
    weights = [ind.weight for ind in population]
    n_female = count(ind -> ind.sex == :female, population)

    # Race counts
    race_counts = Dict{Symbol, Int}()
    for ind in population
        race_counts[ind.race] = get(race_counts, ind.race, 0) + 1
    end

    # Disease severity counts
    severity_counts = Dict{Symbol, Int}()
    has_disease = any(ind -> ind.disease_severity !== nothing, population)
    if has_disease
        for ind in population
            if ind.disease_severity !== nothing
                severity_counts[ind.disease_severity] = get(severity_counts, ind.disease_severity, 0) + 1
            end
        end
    end

    summary = Dict{Symbol, Any}(
        :n => n,
        :age_mean => sum(ages) / n,
        :age_sd => sqrt(sum((ages .- sum(ages)/n).^2) / (n - 1)),
        :age_range => (minimum(ages), maximum(ages)),
        :weight_mean => sum(weights) / n,
        :weight_sd => sqrt(sum((weights .- sum(weights)/n).^2) / (n - 1)),
        :weight_range => (minimum(weights), maximum(weights)),
        :female_proportion => n_female / n,
        :race_distribution => Dict(k => v/n for (k,v) in race_counts)
    )

    if has_disease
        biomarkers = [ind.baseline_biomarker for ind in population if ind.baseline_biomarker !== nothing]
        if !isempty(biomarkers)
            summary[:biomarker_mean] = sum(biomarkers) / length(biomarkers)
            summary[:biomarker_sd] = sqrt(sum((biomarkers .- summary[:biomarker_mean]).^2) / (length(biomarkers) - 1))
        end
        summary[:severity_distribution] = Dict(k => v/n for (k,v) in severity_counts)
    end

    return summary
end
